nixos/modules/services/databases/cockroachdb.nix

```diff
@@ -0,0 +1,221 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.cockroachdb;
+  crdb = cfg.package;
+
+  escape    = builtins.replaceStrings ["%"] ["%%"];
+  ifNotNull = v: s: optionalString (!isNull v) s;
+
+  startupCommand = lib.concatStringsSep " "
+    [ # Basic startup
+      "${crdb}/bin/cockroach start"
+      "--logtostderr"
+      "--store=${cfg.dataDir}"
+      (ifNotNull cfg.locality "--locality='${cfg.locality}'")
+
+      # WebUI settings
+      "--http-addr='${cfg.http.address}:${toString cfg.http.port}'"
+
+      # Cluster listen address
+      "--listen-addr='${cfg.listen.address}:${toString cfg.listen.port}'"
+
+      # Cluster configuration
+      (ifNotNull cfg.join "--join=${cfg.join}")
+
+      # Cache and memory settings. Must be escaped.
+      "--cache='${escape cfg.cache}'"
+      "--max-sql-memory='${escape cfg.maxSqlMemory}'"
+
+      # Certificate/security settings.
+      (if cfg.insecure then "--insecure" else "--certs-dir=${cfg.certsDir}")
+    ];
+
+    addressOption = descr: defaultPort: {
+      address = mkOption {
+        type = types.str;
+        default = "localhost";
+        description = "Address to bind to for ${descr}";
+      };
+
+      port = mkOption {
+        type = types.int;
+        default = defaultPort;
+        description = "Port to bind to for ${descr}";
+      };
+    };
+in
+
+{
+  options = {
+    services.cockroachdb = {
+      enable = mkEnableOption "CockroachDB Server";
+
+      listen = addressOption "intra-cluster communication" 26257;
+
+      http = addressOption "http-based Admin UI" 8080;
+
+      locality = mkOption {
+        type = types.nullOr types.str;
```

Comment:
These strings need to be topographically ordered (from "least specific" qualifier to "most specific"), and it's *very* unclear to me what happens when you merge things and how the ordering works out.

Besides, individual modules cannot possibly know the datacenter/system the database is deployed on. Anyone setting the `locality` option except for the actual `configuration.nix` declaration *in question* is doing something wrong, AFAICS.

Reply:
> and it's very unclear to me what happens when you merge things from two different modules and how the ordering works out.

This is controlled via [mkBefore/mkAfter/mkOrder](https://nixos.org/nixos/manual/#sec-modularity) based on a priority value, so if you want to append a locality you'd just do `locality = mkAfter "foobar"`.

But okay, if you want to prevent that from being possible (except via overrides), then it makes sense to leave it at `str`.

> Besides, individual generic modules cannot possibly know the datacenter/system the database is deployed on

Maybe generic modules might not know, but let's say you have a module that defines these values by discovery of other nodes, eg. [this](https://github.com/headcounter/deployment/blob/c74d9b63a4aa7c0c91b6f0aca1c72b6db3e91b1b/hydra.nix#L84-L93) comes into my mind, where the build slaves for Hydra are discovered using the nodes attribute.

Reply:
> Maybe generic modules might not know, but let's say you have a module that defines these values by discovery of other nodes, eg. this comes into my mind, where the build slaves for Hydra are discovered using the nodes attribute.

That use case works fine, unless I'm missing something -- just set the `locality` setting based on the attribute values you produce. After all, *you* are writing the code to instantiate `services.cockroachdb`, so just do it right there. You're done! The real question is, *why do you want to set locality in two places, in two modules*? Locality is really a cluster-level concept, and the keys/values must be the same for every node in the cluster. It's not something generic modules can (or should) have an interest in, really. Furthermore CockroachDB is a "global service" for NixOS, like most services -- there can't be more than one of it at a time (though it may serve multiple uses).

So under the assumption that one NixOS instance = one CockroachDB node, there is really *only* one global possible locality configuration for any usage of `services.cockroachdb`. And there's no generic module that can know how to configure that. Any module that *does* set it has to be done with your specific deployment in mind. In your scenario, *you're already writing code* to instantiate the module.

The use case that won't work here is if you want to import module `a.nix` and `b.nix` and have them both, somehow, set locality information (specifically locality: they can set other mergeable options just fine). For example, `a.nix` could say `planet=earth` and `b.nix` could say `continent=us`, and they would like to "share" the CockroachDB service, and enable it themselves each. This seems like a misdesign. First off every locality setting must be consistent across all nodes, so every node has to now have `a.nix` and `b.nix` instantiated, or have it replicated (which is a weird workaround, say, if you don't want to enable `b.nix`, but now you have to duplicate the options). Second, if they're going to share a database, you probably want to configure it *independently* of either module, in a single place -- and you can fix that yourself!

(This is sort of a bigger problem with the module system in general and also affects things like PostgreSQL. A lot of things try to offer extensible attributes through merging so that one service can perform multiple roles -- it can be 'reused' through other services -- but it often hits lots of fiddly edge cases because NixOS modules are a *global* concept -- for example, two services can configure/demand mutually exclusive things from a service. In an ideal world, it would be possible to instantiate any module at an arbitrary number of call-sites and have multiple copies coexist, of course. I'm sure you know all this, though)

In short, I think any generic module can't (and should not) have any interest in the locality information, and any situation where you *do* want it should be done by a instantiation of it, per instance, in a single place. I'm having a very hard time coming up with use cases where this doesn't work out?

Reply:
And, on this note, what *would* be cool actually is if I could somehow make `locality` a *non-mergeable* attribute set.

Reply:
Okay, agreed, my example wasn't a particular good one, so maybe it would make sense to wait until the first user comes along and has a use-case for this. 

If you want to make the attribute set non-mergeable, you can just use the `mergeOneOption` merge function, like this:

```nix
{ lib, ... }:

let
  inherit (lib) types;
in {
  imports = [
    { foo.bar = "xxx"; }
    { foo.xxx = "yyy"; }
  ];

  options.foo = lib.mkOption {
    type = (types.attrsOf types.str) // {
      merge = lib.mergeOneOption;
    };
    default = {};
    description = "foo";
  };
}
```

Output of evaluation:

```sh-session
$ nix-instantiate --eval '<nixpkgs/nixos>' --arg configuration /path/to/above/file.nix -A config.foo
error: The unique option `foo' is defined multiple times, in `/home/aszlig/mergey.nix' and `/home/aszlig/mergey.nix'.
(use '--show-trace' to show detailed location information)
$
```

Reply:
Damn, just missed this comment! I just simplified things a bit in 2a2255409265c070d8a3d9a00172452a468aa1ff based on this. I'll submit a follow up with this attribute change, too, since a typed version is much nicer.